### PR TITLE
Adding Promise support for "getHost".

### PR DESCRIPTION
**Pull request by OronNadiv**
@monkpow @villadora 
This PR adds support for getting hostname asynchronously using Promises.  It solves #284.

### CODE REVIEW

File: test/decorateRequest.js [hunk 1]
```diff
@@ -65,7 +65,7 @@ describe('decorateRequest', function() {
         it('returns err to host application for processing', function(done) {
           var app = express();
 
-          app.use(proxy('/reject-promise', {
+          app.use(proxy('httpbin.org', {
```
Comment (OronNadiv) [inline test/decorateRequest.js]: Unrelated issue.  This test would always fail since host __must__ start with `http(s)`.

**Comment by monkpow** [comment]
@OronNadiv Thanks for this patch.    I'm starting to review now.

File: app/steps/resolveProxyHost.js [hunk 2]
```diff
@@ -2,18 +2,20 @@
 var requestOptions = require('../../lib/requestOptions');
 
 function resolveProxyHost(container) {
-  var parsedHost;
+  var promise;
 
   if (container.options.memoizeHost && container.options.memoizedHost) {
-    parsedHost = container.options.memoizedHost;
+    promise = Promise.resolve(container.options.memoizedHost);
```
Comment (monkpow) [inline app/steps/resolveProxyHost.js]: At first glance, this appears to conflict with the goal of memoizing the host.

File: lib/requestOptions.js [hunk 3]
```diff
@@ -24,29 +24,34 @@ function parseHost(Container) {
...
-  }
-
-  var ishttps = options.https || parsed.protocol === 'https:';
-
-  return {
-    host: parsed.hostname,
-    port: parsed.port || (ishttps ? 443 : 80),
-    module: ishttps ? https : http,
-  };
+  host = (typeof host === 'function') ? host(req) : host;
```
Comment (monkpow) [inline lib/requestOptions.js]: dropped toString() here.  investigate why.

File: test/decorateRequest.js (hunk 1 above)
Comment (monkpow) [reply test/decorateRequest.js]: I think this comment is incorrect.    Test pass today without all starting with http.     Current code assumes http if http/https is not provided.

File: lib/requestOptions.js (hunk 3 above)
Comment (OronNadiv) [reply lib/requestOptions.js]: I don't think the `toString()` is needed.  Anyways, I moved it to line [35](https://github.com/villadora/express-http-proxy/pull/285/files#diff-66ad66b7914d1efd19723774d3d91cd8R35) to support backward compatibility.

File: test/decorateRequest.js (hunk 1 above)
Comment (OronNadiv) [reply test/decorateRequest.js]: 👍

File: app/steps/resolveProxyHost.js (hunk 2 above)
Comment (OronNadiv) [reply app/steps/resolveProxyHost.js]: I have a case where an external server defines which users should go to server A and which should go to server B.  In that case, every incoming request would query that external server.  I plan to cache the results (MRU + expiration), but I still need an async support to determine the target host url.  Does it make sense?

**Comment by OronNadiv** [comment]
Thanks for reviewing it @monkpow.  I answered a few of your questions.

**Pull request closed without merging**